File: starter_app/consts/permission.py
from ..utils.enum import KV, SimpleEnum


class Permission(SimpleEnum):
    ## Top level

    # logicunit
    manage_logicunit = KV

    ## System

    # user
    read_user = KV
    manage_user = KV

    # group
    read_group = KV
    manage_group = KV


class PermissionCategory(SimpleEnum):
    operation = KV
    system = KV


permissions_by_category = {
    PermissionCategory.operation: [
        # scene
        Permission.manage_logicunit,
    ],
    PermissionCategory.system: [
        # user
        Permission.read_user,
        Permission.manage_user,
        # group
        Permission.read_group,
        Permission.manage_group,
    ],
}


def to_ts():
    from ..utils.to_ts_lib import echo_dict, echo_enum, echo_enum_list_dict

    echo_enum(Permission)

    echo_enum(PermissionCategory)

    echo_enum_list_dict('permissions_by_category', permissions_by_category, 'Permission', 'key: string', 'string[]')

File: test/frontend/consts/permission.ts
export const Permission = {
  manage_logicunit: "manage_logicunit",
  read_user: "read_user",
  manage_user: "manage_user",
  read_group: "read_group",
  manage_group: "manage_group",
} as const;
export const Permission_keys = Object.keys(Permission);
export const Permission_values = Object.values(Permission);

export const PermissionCategory = {
  operation: "operation",
  system: "system",
} as const;
export const PermissionCategory_keys = Object.keys(PermissionCategory);
export const PermissionCategory_values = Object.values(PermissionCategory);

export const permissions_by_category: {[key: string]: string[]} = {
  operation: [Permission.manage_logicunit],
  system: [Permission.read_user, Permission.manage_user, Permission.read_group, Permission.manage_group],
}

